### ISSUE TITLE
RFE: please parametrize old_signing_keys in roles/matrix-synapse/templates/synapse/homeserver.yaml.j2

### ISSUE THREAD

**Issue by pcfe** [spantaleev/matrix-docker-ansible-deploy#738]
If a user needs to publish previously used keys, it would be useful if `old_signing_keys:` in `roles/matrix-synapse/templates/synapse/homeserver.yaml.j2` could be filled via entries in `vars.yml`.

Reasoning; I did a disaster recovery test but failed to copy back Synapse's signing key, in `/matrix/synapse/config`, before starting the services for the first time. So now I need to perpetually publish the old pubkey (I do have a backup of it)
c.f. [here](https://matrix.to/#/!ehXvUhWNASUkSLvAGP:matrix.org/$NkcNwoWTHXeeCuyJ99I5bmHb8dDM6yfloY39-b5C610?via=matrix.org&via=blob.cat&via=tchncs.de) and [here](https://matrix.to/#/!PukFFdIcHgtaaHZflT:devture.com/$16065942874rxNnP:pcfe.net?via=matrix.org&via=envs.net&via=linuxdelta.com)

**Comment by spantaleev** [comment]
One could always use `matrix_synapse_configuration_extension_yaml`, but we may provide a dedicated variable as well.

**Comment by pcfe** [comment]
yeah, works for now with `matrix_synapse_configuration_extension_yaml` block.
This RFE is to track easing that in the future.

**Comment by xthursdayx** [comment]
@pcfe can share the `matrix_synapse_configuration_extension_yaml` block you used (redacted as necessary)? I have the same problem and have my old signing key, but am not sure what the line in my vars.yml should be. Thanks!

**Comment by spantaleev** [comment]
```yaml
matrix_synapse_configuration_extension_yaml: |
  old_signing_keys:
    "ed25519:a_ABCD": { key: "ABCDEfGHIJKLMNOP", expired_ts: 123456789123 }
```

The above example was for a `*.signing.key` file that contained data like this:

```
ed25519 a_ABCD ABCDEfGHIJKLMNOP
```

Make sure to pick some expiration time from https://currentmillis.com/

At least, that's how I understand it.

**Comment by spantaleev** [comment]
Oh, looks like this is wrong.

I thought that the 3rd thing you see in the `*.signing.key` file was what was necessary (`old_signing_keys` calls for the base64 encoded key), but it's not.

Calling `/usr/local/bin/export_signing_key` on my key does yield a different value.

---------------

I think the correct way is this:

You can put your old key in `/matrix/synapse/config/old.key`, then call: `docker exec matrix-synapse /usr/local/bin/export_signing_key -x /matrix/synapse/config/old.key` and paste the generated value here:

```yaml
matrix_synapse_configuration_extension_yaml: |
  old_signing_keys:
    HERE
```

**Comment by xthursdayx** [comment]
Thanks a lot @spantaleev !

Just to note, the command `docker exec matrix-synapse /usr/local/bin/export_signing_key -x /matrix/synapse/config/old.key` didn't work. After entering the matrix-synapse container I found that the `/matrix/synapse/config` is actually located at `/data` within the container, so the command `docker exec matrix-synapse /usr/local/bin/export_signing_key -x /data/old.key` did the job.

However, unfortunately, I'm still getting the same error when I try to view public rooms, despite adding the `old_signing_keys` lines to my vars.yml.

**Comment by aaronraimist** [comment]
Did you run the `setup-all` and `start` tags after adding it to `vars.yml`?

Does the old key appear in https://matrix.domain:8448/_matrix/key/v2/server?

**Comment by xthursdayx** [comment]
Yes, I've run `setup-all` and `start` a handful of time. This is what I see at https://matrix.domain:8448/_matrix/key/v2/server:
```
{"old_verify_keys":{"ed25519:a_NZUH":{"expired_ts":1611104471398,"key":"fnLopZRueWibTafmkd/P5R4l+Q8+q46XYeTCFFmDAmg"}},"server_name":"my.domain","signatures":{"my.domain":{"ed25519:a_wGBG":"sQ1HSHO34ONBmUnfwo1Ldq2e/kOary8J324HVyta3oeUvA0jGP7m9kIgT0hiQkIWoIxdXAJ6259OHI3LkgDQBg"}},"tls_fingerprints":[{"sha256":"yqI95EQt3CoGmq2nKn7Y2Rqp0dN49ZAr6urVGIwTJw4"}],"valid_until_ts":1611170242492,"verify_keys":{"ed25519:a_wGBG":{"key":"EQoCPwtKc7dciMQN/8LDJcG9YIsxWQKLoFPovMXVQbo"}}}
```

The old key shown does match the output of `docker exec matrix-synapse /usr/local/bin/export_signing_key -x /data/old.key`

**Comment by pcfe** [comment]
> @pcfe can share the `matrix_synapse_configuration_extension_yaml` block you used (redacted as necessary)? I have the same problem and have my old signing key, but am not sure what the line in my vars.yml should be. Thanks!

apologies, I had forgotten to do this in a timely fashion.

Here's what I use:

```yaml
# Old keys
matrix_synapse_configuration_extension_yaml: |
  old_signing_keys:
    "ed25519:foo": { key: "bar…baz", expired_ts: 1606662940663 }
```

**Comment by luixxiul** [comment]
Closing as resolved.
